crates/ruff/src/rules/pycodestyle/rules/logical_lines/mod.rs

```diff
@@ -455,6 +456,10 @@ impl LogicalLinesBuilder {
     fn push_token(&mut self, kind: TokenKind, range: TextRange) {
         let line = &mut self.current_line;
 
+        if !matches!(kind, TokenKind::NonLogicalNewline) {
+            line.non_empty = true;
+        }
```

Comment:
Do you think it would work if we instead test inside of `finish_line` if all `tokens` are `NonLogicalNewLine`? That would have the benefit that it only runs once per line, doesn't require more state, and doesn't introduce more branching in this hot path (at the cost that, in the worst case, we have to iterate over all line tokens). We could optimize this further by providing a `finish_non_logical_line` and only add the check there (calls into the `finish_line`